0010-B-object-storage-building-block.md

```diff
@@ -0,0 +1,264 @@
+# Object storage building block
+
+* Author(s): @ItalyPaleAle
+* State: Draft
+* Updated: 2022-12-14
+* Original proposals / discussions: dapr/dapr#4808, dapr/dapr#4934
+
+## Overview
+
+This is a design proposal for a new "object storage" building block which allows Dapr users to store and retrieve unstructured data of arbitrary size.
+
+We define **objects** as sequences of unstructured data, that should be assumed to be possibly large (many MBs or even GBs in size) and binary. Examples include images and videos, Office documents, etc.
+
+## Implementation Details
+
+### Building block interface
+
+The proposal involves creating a new `objectstore` building block with the following interface:
+
+```go
+type ObjectStore interface {
+    // GetObject retrieves an object with name key and writes it to the out stream.
+    // If the object doesn't exist, err will be ErrNotFound.
+    // Return value tagOut contains an etag or a similar concurrency-control tag.
+    GetObject(ctx context.Context, key string, out io.Writer) (md ObjectMetadata, tagOut any, err error)
+    // SetObject stores an object with name key, reading data from the in stream.
+    // Parameter tag is optional and can be used to pass an etag or any similar concurrency-control tag.
+    // In case of conflict (e.g. etag mismatch), the method returns ErrConflict.
+    SetObject(ctx context.Context, key string, in io.Reader, tag any, md ObjectMetadata) (tagOut any, err error)
+    // DeleteObject deletes an object with name key.
+    // Parameter tag is optional and can be used to pass an etag or any similar concurrency-control tag.
+    // In case of conflict (e.g. etag mismatch), the method returns ErrConflict.
+    DeleteObject(ctx context.Context, key string, tag any) (err error)
+}
+
+// Metadata associated with an object.
+type ObjectMetadata map[string]string
+
+// Error constants:
+var ErrNotFound = errors.New("...")
+var ErrConflict = errors.New("...")
+```
+
+The `key` parameter is the name of the object. Unlike with state stores, there are no key prefixes when operating with objects: applications have full access to all data in the storage service.
+
+Each object can have metadata associated to it, which is normally passed to clients as headers (more on that below). In particular, `Content-Type` is yet another metadata key, treated no differently from anything else.
+
+> Option: we could support auto-detection of the content type if users pass `"auto"` as value, using a library such as [ItalyPaleAle/file-type-stream-go](https://github.com/ItalyPaleAle/file-type-stream-go) (which I'd be very happy to transfer to the Dapr org).
+
+The only exception is the `tag`, which is passed separately from the metadata object. The reason is that in the `SetObject` method, the tag is both an (optional) parameter and a return value. Note however that support for tags / ETags, is not a requirement if the underlying storage service doesn't support it.
```

Comment:
DO we really need to have etag support? Which services support etag for blob storage?

Reply:
I think it's pretty important to have ETag support. AFAIK, all major object storage services support etags: Azure Blob Storage, S3 (and every S3-compatbile), etc.

Reply:
Sorry, getting back to this issue: but I don't think S3 has any support for concurrent writes.

It might be worth pointing out that not all object stores support concurrency-control. [S3 doesn't](https://stackoverflow.com/questions/14594924/s3-concurrent-writes).  GCP blob-storage does.